### DISCUSSION TITLE
New component utility functions

### DISCUSSION THREAD

**Discussion by NicholasBoll** [Workday/canvas-kit#1578] [Ideas]
## `createComponent`

`createComponent` was created as a lightweight wrapper around `React.forwardRef` that added handling for the `as` prop. This utility function's largest value-add is how it handles polymorphic types. For example:

```tsx
interface MyComponentProps {
  variant: 'large' | 'small';
  children?: React.ReactNode;
}

const MyComponent = createComponent('div')({
  Component: (props: MyComponentProps, ref, Element) {
    ref // React.Ref<HTMLDivElement>
    Element // 'div'

    return <Element ref={ref} {...props} />
  }
});

// MyComponent extends the HTMLDivElement attribute interface
const temp1 = <MyComponent onClick={e => console.log(e)} />
// e is React.MouseEvent<HTMLDivElement, MouseEvent>

const temp2 = <MyComponent as="aside" onClick={e => console.log(e)} />
// e is React.MouseEvent<HTMLElement, MouseEvent>
```

This polymorphism is also why we stopped exporting the HTML attribute interface on component props. We introduced the `ExtractProps` instead:

```tsx
type temp1 = ExtractProps<typeof MyComponent> // MyComponentProps & React.HTMLAttributes<HTMLDivElement>
type temp2 = ExtractProps<typeof MyComponent, never> // MyComponentProps
type temp3 = ExtractProps<typeof MyComponent, 'aside'> // MyComponentProps & React.HTMLAttributes<HTMLElement>
```

`ExtractProps` knows what the default HTML interface of a component is and also allows you to override it if needed if your component uses an `as` and is not polymorphic itself.

## Common Patterns

Compound components with models all follow a common pattern which meant common boilerplate. This is more noticeable with recent work to remove boilerplate from models. In compound components with a model, there are 2 types: a container component and a subcomponent. Container components always create a model's context and provide it to the React render tree. If a `model` is passed to a container, it will use it - if not, the container will create a model. A subcomponent, by contrast, will take a passed `model` from props, or extract it from React's context.

This led to boilerplate and also an issue with circular imports. A container component imports all subcomponents and exposes them using dot notation:

```tsx
// ContainerComponent.tsx
import { SubComponent } from './SubComponent'

export const ModelContext = React.createContext({})

export const ContainerComponent = createComponent()({
  /// ....
  subComponents: {
    SubComponent,
  }
});

// SubComponent.tsx
import {ModelContext} from './ContainerComponent'

// ...
```

Webpack 4 handles this circular import, but not all bundlers do.

## Container components

Let's look at the boilerplate common to all container components:

```tsx
import React from 'react';
import {createComponent} from '@workday/canvas-kit-react/common';
import{MyModel, MyModelConfig, useMyModel} from './model';
import {SubComponent} from './SubComponent';
import {useContainerComponent} from './hooks';

export const ModelContext = React.createContext({});

interface ContainerComponentProps extends MyModelConfig {
  children?: React.ReactNode;
  model?: MyModel;
}

export const ContainerComponent = createComponent('div')({
  Component({model, modelConfig1, modelConfig2, ...props}: ContainerComponentProps, ref, Element) {
    // We use `localModel` instead of `model` to avoid the shadow variable ESLint rule
    const localModel = useDefaultModel(model, {modelConfig1, modelConfig2}, useMyModel);
    // useContainerComponent is a standard elemProps hook. The implementation is not important
    const elemProps = useContainerComponent(localModel, ref, props);

    return (
      <ModelContext.Provider value={localModel}>
        <Element ref={ref} {...elemProps} />
      </ModelContext.Provider>
    );
  },
  subComponents: {
    SubComponent,
  },
});
```

We'll notice a few oddities.
- We always have to include `children` and `model` in the prop interface even though all container components have these props. This allows room for error.
- The Container component makes the context. The problem is the container component needs the subcomponent and the subcomponent needs the container component's context
- The container component has to know all the config options of the model so it can extract them out and hand them to `useDefaultModel` so that the `elemProps` object will only be props to be handed to the `<Element>` component
- `elemProps` could contain a `ref` if `useContainerComponent` needed to create a forked ref
- The new paradigm for creating model hooks no longer exports `MyModel` or `MyModelConfig` types!

Here's what it will have to look like with the new models:
```diff
 import React from 'react';
import {createComponent} from '@workday/canvas-kit-react/common';
-import {MyModel, MyModelConfig, useMyModel} from './model';
+import {useMyModel} from './model';
import {SubComponent} from './SubComponent';
 
-const ModelContext = React.createContext({});
+const ModelContext = useMyModel.Context;
 
+type MyModelConfig = Partial<typeof useMyModel.defaultConfig> & typeof useMyModel.requiredConfig;
 interface ContainerComponentProps extends MyModelConfig {
   children?: React.ReactNode;
-  model?: MyModel;
+  model?: ReturnType<typeof useMyModel>;
 }
 
 export const ContainerComponent = createComponent('div')({
   Component({model, modelConfig1, modelConfig2, ...props}: ContainerComponentProps, ref, Element) {
     // We use `localModel` instead of `model` to avoid the shadow variable ESLint rule
     const localModel = useDefaultModel(model, {modelConfig1, modelConfig2}, useMyModel);
     // useContainerComponent is a standard elemProps hook. The implementation is not important
     const elemProps = useContainerComponent(localModel, ref, props);

    return (
      <ModelContext.Provider value={localModel}>
        <Element ref={ref} {...elemProps} />
      </ModelContext.Provider>
    );
   },
   subComponents: {
     SubComponent,
```

The model no longer exports the types we need to create a container component, so we have to recreate them. This is a pain. We've added more information to the `useMyModel` hook - `useMyModel.Context`, `useMyModel.defaultConfig`, and `useMyModel.requiredConfig`, we can make a container factory function to replace `createComponent` that takes in a model hook and adds all the extra types for us:

### Introducing `createContainer`

```tsx
interface ContainerComponentProps {}

export const ContainerComponent = createContainer('div')({
  modelHook: useMyModel,
  elemPropsHook: useContainerComponent,
  subComponents: {
    SubComponent,
  },
})<ContainerComponentProps>((elemProps, Element) => {
  return <Element {...elemProps} />;
});
```

Here are the benefits:
- `model` and `children` are no longer needed in the `ContainerComponentProps` interface. The prop interface can focus on props specific to this component
- We no longer need to worry about proper `ref` handling
- We no longer need to worry about providing the model context
- The `modelHook` takes care of model value and types, we no longer need to create a `localModel` variable
- The `elemPropsHook` allows us to keep the render function focused on only rendering. We no longer need the `elemProps`/`props` variables
- There's a 3rd parameter called `model` that provides the model if we need it, but this example does not

The external interface of `ContainerComponent` includes the optional  `model` prop and all optional model config automatically without us having to type it!

```tsx
type foo = ExtractProps<typeof ContainerComponent> // {model?: MyModel, children?: React.ReactNode} & ContainerComponentProps
```

The funky part left in this example is the `<ContainerComponentProps>((elemProps, Element) => {})`. We do this because `elemProps` contains additional props added by the `elemPropsHook`, including `children`. This syntax is a little foreign at first, but it allows us to maximize Typescript's inference while minimizing how much Typescript code we need.

## Subcomponents

Subcomponents had similar, but different boilerplate to container components. A subcomponent would receive an optional `model` and use that model or grab from context. To avoid `rules-of-hooks` ESLint errors, we used a `useModelContext` hook that did this model handling for us.

Using `createComponent` with current models
```tsx
interface SubComponentProps {
  children?: React.ReactNode;
  model?: ReturnType<typeof useMyModel>;
}

export const SubComponent = createComponent('div')({
  Component({model, ...props}: SubComponentProps, ref, Element) {
    const localModel = useModelContext(useMyModel.Context, model);
    const elemProps = useSubComponent(localModel, ref);
    return <Element {...elemProps} />;
  },
});
```

### Introducing `createSubcomponent`

```tsx
interface SubComponentProps {}

export const SubComponent = createSubcomponent('div')({
  modelHook: useMyModel,
  elemPropsHook: useSubComponent,
})<SubComponentProps>((elemProps, Element) => {
  return <Element {...elemProps} />;
})
```

`createSubcomponent` takes care of the model finagling we had to do as well as the prop dance. The `SubComponentProps` can focus on any unique props the component takes that effects rendering. Like `createContainer`, an optional 3rd parameter is the `model` in case your render function needs access to it. Most likely this is because your render function renders more than a single element, so the `elemPropsHook` doesn't quite meet your needs.

### When to use `createComponent`

`createComponent` is not going away. It is still useful if you want to create a polymorphic component that forwards a `ref` and `elemProps`. Use it if your component doesn't use models. For example, our button components don't have models and still use `createComponent`. The new utility functions are useful in creating compound components with models that match our specifications without extra boilerplate.

**Comment by mannycarrera4** [comment]
This is an awesome write up! A few notes:

- I would highlight or create a new title something of the likes of `New way of createComponent and it's benefits` you small text that says `Here's what it will have to look like with the new models:` and it kind of gets lost.
- Also in your examples `Let's look at the boilerplate common to all container components:` There's some valuable information knowing what `localModel` and `useContainerComponent` are doing since this changes in the new format. Is there a way to either add inline comments in that example or explain what it's doing?
- After your example after this sentence: `we can make a container factory function to replace createComponent that takes in a model hook and adds all the extra types for us` i think you should break out all the changes and goodies into bullet points rather than describing it in a paragraph
- Same thing for `createSubComponent` list out in bullet points what it's doing, it's removing the need to add `model` as a prop
- Can you go more in depth as to what is `elemPropsHooks` i think you mention it removed a few extra lines but what is the abstraction?
- Last piece i would maybe mention is, can I still use `createComponent` not all component will need a model right? So there's three factory functions, `createComponent`, `createContainer` and `createSubcomponent` maybe a small section describing when to use each one. I think the logic is, if you have a model and need to share with across other components `createContainer` for your wrapper component and `createSubcomponent` for any sub components that need that model, otherwise, `createComponent` for a use case without a model?

**Comment by NicholasBoll** [comment]
> Also in your examples Let's look at the boilerplate common to all container components: There's some valuable information knowing what localModel and useContainerComponent are doing since this changes in the new format. Is there a way to either add inline comments in that example or explain what it's doing?

I normally like inline comments to help narrate the point, but I didn't want to distract from the difference with comments. `localModel` is a rename of `model` to get around ESLint rules, and `useContainerComponent` is an elemProps hook where the implementation isn't important to the problem
